hugo/content/posts/nix-bazel-crosspiling.md

```diff
@@ -0,0 +1,129 @@
+---
+title: "Nix + Bazel Crosspiling"
+date: 2023-05-24T18:00:00-07:00
+author: James Leitch
+tags:
+  - nix
+  - nixpkgs
+  - bazel
+  - bzlmod
+  - crosspile
+  - crosspiling
+  - toolchain
+  - platform
+---
+
+**Code: <https://github.com/ConsumingChaos/consumingchaos.github.io/tree/master/examples/nix-bazel-crosspiling>**
+
+# Motivation
+
+For years now, I've had "learn and set up crosspiling" as personal backlog item. From my career observations, there is often a lot of developer pain and efficiency lost when a project is targeting multiple platforms. Some examples that come to mind:
+
+- Having to switch developer machines and synchronize code across them.
+- Having to rig up together some form of individual remote compilation and teach everyone how to use it (for example, having Windows sync the repo to Mac, run the build on Mac, and report back the results).
+- Relying on continuous integration to build non-host platforms and not having coverage during normal developer iteration.
+
+Conversely, I see some major upsides to crosspiling:
+
+- Running "build all" can actually build **all**.
+- Developer environments can be standardized to a single development environment (like a Linux [DevServer]({{< ref "/posts/nixos-devserver.md" >}})).
+- Automation infrastructure (like continuous integration) can also be standardized to the same single development environment.
+
+With these motivations in mind, the question becomes, how do I actually set this up? From the reading and searching I've done, I had a hard time finding a concise and clear example of crosspiling "from scratch". This was particularly evident with my preferred choice of package management and build system, Nix and Bazel respectively. Hopefully this post can help serve as such an example.
+
+For background information and motivation on why Nix + Bazel, I recommend reading <https://www.tweag.io/blog/2018-03-15-bazel-nix/>.
+
+# Nix
+
+Nix is the package manager providing all the tools and SDKs needed to crosspile code. To pull this off, my [Nix Flake](https://github.com/ConsumingChaos/consumingchaos.github.io/tree/master/examples/nix-bazel-crosspiling/flake.nix) exposes two things, a `devShell` (making tools available in the terminal and VSCode) and a set `bazel` that contains derivations that get exposed to Bazel via [rules_nixpkgs](https://github.com/tweag/rules_nixpkgs). Keeping everything in the same Nix Flake lets me ensure there is only one set of tools and SDKs and that they're shared between the developer environment and Bazel.
```

Comment:
You probably want a colon here:

"My Nix Flake exposes two things:"

I would also clean up this sentence:

"Keeping everything in the same Nix Flake lets me ensure there is only one set of tools and SDKs, shared between the developer environment and Bazel."